collector/agentPoller.go

```diff
@@ -15,451 +15,433 @@
-		return nil, err
+		msg.Dimensions.AgentID = in.agentState.ID
+		msg.Dimensions.ClusterID = "" // TODO(roger) need to get this from the master
+		msg.Dimensions.ContainerID = c.ContainerID
+		msg.Dimensions.FrameworkID = c.FrameworkID
+		msg.Dimensions.FrameworkName = fi.Name
+		msg.Dimensions.FrameworkRole = fi.Role
+		msg.Dimensions.FrameworkPrincipal = fi.Principal
+		msg.Dimensions.ExecutorID = c.ExecutorID
+		msg.Dimensions.ContainerID = c.ContainerID
+		msg.Dimensions.Labels = getLabelsByContainerID(c.ContainerID, in.agentState.Frameworks)
+
+		out = append(out, msg)
 	}
 
-	// state["frameworks"][N]["id"] (framework_id)
-	// => state["frameworks"][N]["name"] (framework_name)
-	frameworkNames := make(map[string]string, len(frameworks))
-
-	// state["frameworks"][N]["executors"][M]["id"] (executor_id)
-	// => state["frameworks"][N]["executors"][M]["labels"][L(MARATHON_APP_ID)]["value"] (application_name)
-	executorAppNames := make(map[string]string, 0)
+	return out
+}
 
+// getFrameworkInfoByFrameworkID returns the FrameworkInfo struct given its ID.
+// For example: "5349f49b-68b3-4638-aab2-fc4ec845f993-0000" => "marathon"
+func getFrameworkInfoByFrameworkID(frameworkID string, frameworks []frameworkInfo) (frameworkInfo, bool) {
 	for _, framework := range frameworks {
-		frameworkID, err := framework.GetString("id")
-		if err != nil {
-			stats <- statsd.MakeEvent(statsd.AgentQueryBadData)
-			return nil, err
-		}
-		frameworkName, err := framework.GetString("name")
-		if err != nil {
-			stats <- statsd.MakeEvent(statsd.AgentQueryBadData)
-			return nil, err
-		}
-		frameworkNames[frameworkID] = frameworkName
-
-		executors, err := framework.GetObjectArray("executors")
-		if err != nil {
-			stats <- statsd.MakeEvent(statsd.AgentQueryBadData)
-			return nil, err
-		}
-		for _, executor := range executors {
-			executorID, err := executor.GetString("id")
-			if err != nil {
-				stats <- statsd.MakeEvent(statsd.AgentQueryBadData)
-				return nil, err
-			}
-			labels, err := executor.GetObjectArray("labels")
-			if err != nil {
-				// ignore this failure: labels are often missing when it's not a marathon app.
-				continue
-			}
-			// check for marathon app id. if present, store as application name:
-			for _, label := range labels {
-				labelKey, err := label.GetString("key")
-				if err != nil {
-					stats <- statsd.MakeEvent(statsd.AgentQueryBadData)
-					return nil, err
-				}
-				_, ok := marathonAppIDLabelKeys[labelKey]
-				if ok {
-					labelValue, err := label.GetString("value")
-					if err != nil {
-						stats <- statsd.MakeEvent(statsd.AgentQueryBadData)
-						return nil, err
-					}
-					executorAppNames[executorID] = strings.TrimLeft(labelValue, "/")
-				}
-			}
+		if framework.ID == frameworkID {
+			return framework, true
 		}
 	}
-
-	return &AgentState{
-		agentID:          agentID,
-		frameworkNames:   frameworkNames,
-		executorAppNames: executorAppNames}, nil
+	return frameworkInfo{}, false
```

Comment:
Init struct to var early and often! This can come back to bite you will nil values later. I always try to `var thing myStruct{}` if there is no data in it. Then you can return `thing` instead of init'ing a new struct later.